### PR TITLE
acmeserver: add policy field to define allow/deny rules

### PR DESCRIPTION
**Pull request by mohammed90**
TODO:
- [x] Tests
- [x] Docs
- [x] Caddyfile support

To be merged after #5794

### CODE REVIEW

**Comment by mholt** [comment]
Cool, this looks great so far. Again, no real strong feedback yet -- it's similar to how I'd probably do it if I had need of this.

**Comment by hslatman** [comment]
@mohammed90 you could drop the URI domains from this. They're only used for when there are URIs in a certificate (request), and those are not a valid ACME identifier type. 

Let me know if you hit issues with the policy configuration / evaluation. We kept it fairly simple intentionally, and sometimes it can be too limiting (e.g. strict equality check of the common name), so happy to hear feedback about it.

**Comment by mohammed90** [comment]
> @mohammed90 you could drop the URI domains from this. They're only used for when there are URIs in a certificate (request), and those are not a valid ACME identifier type.

Thanks for the tip! I'll make the change when I circle back to this PR. Currently the progress on this PR is blocked #5794 since both touch the same file. I'm stuck on testing the ACME server due to my knowledge gap of how to use github.com/mholt/acmez 😶

File: modules/caddypki/acmeserver/policy.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	// to be approved or denied for certificates. Non-CIDR IP addresses
+	// are matched exactly.
+	IPRanges []string `json:"ip_ranges,omitempty"`
+}
+
+// Policy defines the criteria for the ACME server
+// of when to issue a certificate. Refer to the
+// [Certificate Issuance Policy](https://smallstep.com/docs/step-ca/policies/)
+// on Smallstep website for the evaluation criteria.
+type Policy struct {
```
Comment (mholt) [inline modules/caddypki/acmeserver/policy.go:27]: Nit, but move this to the top of the file, it makes it easy to see what this file is about; whereas RuleSet is just a member of Policy.

**Review by mholt — APPROVED** [review]
Awesome, looking good and clean and tested. Just one random thought but nothing of importance. Thank you Mohammed, this is great!

File: modules/caddypki/acmeserver/policy.go (hunk 1 above)
Comment (mohammed90) [reply modules/caddypki/acmeserver/policy.go:12]: Done

**Pull request merged**